### DISCUSSION TITLE
fprime on STM32H7 followup

### DISCUSSION THREAD

**Discussion by jck1044** [nasa/fprime#1933] [Q&A]
Hello,

My name is Jared King and I am a part of the 3UCubed CubeSat team. We are trying to implement F’ on an ARM Cortex M7 MCU – STM32H753II on an evaluation board to emulate our onboard computer.

A couple of weeks ago, a group of us met with Jeffrey L, Kevin O, and Michael S regarding getting fprime working on an STM32H7. To recap our current position, we took the approach of creating the F' build in a linux container using F' utilities and then flashing the build onto the STM board. We are attempting to use FreeRTOS and implement our own OS Abstraction Layer (OSAL). We referred a lot of our implementation to the Melbourne[ Space Program’s repository](https://github.com/MelbourneSpaceProgram/fprime/tree/feat/osal). At the moment, we managed to get an error-free build process for the Fprime, but it seems we can’t get the program running when it’s flashed onto the board. We haven’t been able to identify the cause yet.

The team at JPL gave us loads of information and tips to help us down the correct path. These have not fixed the problem yet, but they are getting us much closer. I will simplify down to three main points:
1. The file being flashed could be too big for the memory available on the STM32.
   - We are flashing to the STM32 with [STLINK](https://github.com/stlink-org/stlink). We found that it return an error when the file is too big to be flashed, so we ruled this out. 
2. Simplify Ref/Top/Main.cpp
   - I was told that essentially the only lines needed were 
   ```
   state = Ref::TopologyState(hostname, port_number);
   Ref::setup(state);
   ```
   - I understand why this is done, but hostname and port_number are still in their initialized state when passed to TopologyState. Could this cause more issues?
   - I was also instructed to change `int main(int argc, char* argv[]) {` to `​​int fsw_main(int argc, char* argv[]) {` so that it does not butt heads with my main method in Os/stm32-bsp/Core/Src/Main.c. I made this change, but I have a few concerns with it. Would I need to change calls to it from any other files? Doesn’t changing the name to something other than “main” make it so it is no longer the main entry point? 
3. Include incremental outputs for debugging purposes.
   - The only successful debugging I was able to do was message(“...”) in CMakeList.txts, which all printed during compilation. I tried multiple methods of outputs in both Os/stm32-bsp/Core/Src/Main.c and Ref/Top/Main.cpp, but I never saw any of them. I tried cout, printf, system(“echo …”), and even calling to execute a file that would make the STM32 flash its LEDs. None of these showed any changes. This was both before and after changing the name of main. Could this mean that it is never reaching this code? 

My repository can be found here: https://github.com/3UCubed/Flight-Software/tree/Jared-Test1/fprime/fprime-feat-osal
Any feedback is appreciated:)

**Comment by thomas-bc** [comment]
Hi Jared!

This was a little while ago so I could be mistaken, but I believe the renaming `main()` was suggested to be done the other way. Keep `main()` in _Ref/Top/Main.cpp_, and rename your main in _Os/stm32-bsp/Core/Src/Main.c_ to something like `fsw_main()`. And from Ref/Top/main.cpp in `main()`, call your `fsw_main()`. This feels less error-prone to me, but maybe your way would work too. 

I tried really quickly to clone and build your repository but got some include errors on FreeRTOS headers when running `fprime-util build` from `Ref/`, are you adding some parameters to that?

**Reply by LeStarch** [reply]
I think the intent was the other way around:

- Os/stm32-bsp/Core/Src/Main.c: entry point of FreeRTOS, should contain `main()` with a call to `fsw_main()`
- Ref/Top/Main.cpp: entry point to F´.  Should define `fsw_main()` to be called from `main()` in FreeRTOS

Make sure the `fsw_main` in `Ref/Main.cpp` is using C linkage so it can be called from your `main.c` FreeRTOS program.

**Reply by thomas-bc** [reply]
Ohh ok my bad. Then maybe all that was missing in [this version of the repo](https://github.com/3UCubed/Flight-Software/tree/6c113e4f68e9e5a571e041283a9dc3d8e830b5f7/fprime/fprime-feat-osal) was the call to fsw_main() from FreeRTOS' `main()`

**Reply by LeStarch** [reply]
Yes.  Likely that call should be made here:

https://github.com/3UCubed/Flight-Software/blob/6c113e4f68e9e5a571e041283a9dc3d8e830b5f7/fprime/fprime-feat-osal/Os/stm32-bsp/Core/Src/main.c#L213

**Reply by jck1044** [reply]
I believe that I was able to call fsw_main() in Ref/Top/Main.cpp from main() in Os/stm32-bsp/Core/Src/main.c. But again, it is hard to be sure when I still have not found a great debugging method.

I believe my next task would be changing the entry point of FreeRTOS, which I am not exactly sure how to do. My most educated guess would be altering Ref/CMakeLists.txt.

Is adding [Main.cpp to SOURCE_FILES in Ref/CMakeLists.txt](https://github.com/nasa/fprime/blob/master/Ref/CMakeLists.txt#L50) where it usually "enters" Main.cpp? Would this need to be removed? Then does a line that points to "Os/stm32-bsp/Core/Src/main.c" need to be inserted?

Thanks again for the continued help

**Comment by jck1044** [comment]
Thank you both for the clarification. I might have fixed the import errors today, but I can't be sure. I made a branch for both options (call to fsw_main from Ref/Top's main<https://github.com/3UCubed/Flight-Software/tree/RefCallsStm32/fprime/fprime-feat-osal> and call to fsw_main from FreeRTOS' main<https://github.com/3UCubed/Flight-Software/tree/Stm32CallsRef/fprime/fprime-feat-osal>). I will continue to work with the latter for now.

**Reply by BillYJT** [reply]
Hi @jck1044 
Glad to see some new progress on this line of investigation!
Here is something that we have learned from our end that you may find useful:

- @LeStarch suggests the inclusion of the startup file in this post: https://github.com/nasa/fprime/discussions/978#discussioncomment-1859811, which seems to define the entry point i.e. _the_ main function.
- For debugging I have had some limited success with setting up and GDB debugger over OpenOCD, following this post: https://elrobotista.com/en/posts/stm32-debug-linux/

**Comment by ebenezerarunkumar** [comment]
@jck1044 @BillYJT Did either of you ever complete a working reference deployment of fprime on an STM board with FreeRTOS?

**Reply by jck1044** [reply]
Hi @EbenezerA99 
Unfortunately, I do not believe that we did. Shortly after this thread, I changed jobs, and work was passed on to others. They ended up changing away from using fprime.

**Reply by codeflight1** [reply]
While not with FreeRTOS, I have been working with getting Fprime to run on the H7, and have a baremetal build working which I can share if that would be helpful.

**Reply by SterlingPeet** [reply]
@EbenezerA99 It might be worth trying to compile fprime for FreeRTOS with the baremetal OSAL.  That path would reduce the friction to get a runnable deployment, at the cost of time services and threading.  I have not tried this for FreeRTOS, so consider the suggestion in that light.

**Reply by ebenezerarunkumar** [reply]
Thanks for the suggestion @SterlingPeet , i was thinking of looking into that. Looks like @codeflight1 has a baremetal build that could be a good starting point... @codeflight1 If you're ok with sharing your deployment, i think it would benefit the community!

**Reply by codeflight1** [reply]
I've uploaded my basic deployment here: https://github.com/codeflight1/fprime-stm32-reference. It's quite messy and missing some key parts (IO components and a hardware rate driver) but should be able to get you up and running. fprime-baremetal currently has a couple cmake problems which I've tried to fix in fprime-community/fprime-baremetal#12 but if that hasn't been merged yet you can pull my [cmake-fixes](https://github.com/codeflight1/fprime-baremetal/tree/cmake-fixes) branch and use that instead of the submodule. I also encountered some weird errors in places where `Fw::ComPacket` is serialized. Not totally sure if this is an STM type configuration error or an fprime problem, but if you encounter it you can just add static casts to `FwPacketDescriptorType` and that should fix it. Depending on your oscillator configuration you may need to update `SystemClock_Config` and `PeriphCommonClock_Config` in `STM32H743/Startup/main.cpp`. You can get these my making an STM32CubeIDE project, setting up your oscillator, and then copying the generated code. If you have any problems or questions feel free to drop them here or create issues over on my repo. 

UPDATE: The CMake fixes (fprime-community/fprime-baremetal#12) and casting errors (#3707) have been fixed
